### PR TITLE
Update to ESLint 4

### PR DESCRIPTION
**Pull request by anselmbradford**
## Removals

- Deprecated ESLint rules.

## Changes

- Upgrades to ESLint 4 and adds new ESLint 3/4 rules.

## Testing

1. `gulp lint && gulp build && gulp test:acceptance`

## Notes

- Majority of code changes are autofix changes, such as indentation.
- The `multiline-comment-style` I wasn't totally set on, but it's an autofix rule so is easy to switch around.
- `globals` needs a value other than null, so commented it out.

### CODE REVIEW

File: cfgov/unprocessed/js/apps/financial-well-being/fwb-questions.js [hunk 1]
```diff
@@ -1,5 +1,3 @@
-'use strict';
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/apps/financial-well-being/fwb-questions.js]: Funny that 4 removes this, I feel like I was constantly adding it in mindlessly.

File: cfgov/unprocessed/js/modules/Analytics.js [hunk 2]
```diff
@@ -66,11 +65,11 @@ var Analytics = {
    * @param {object} dataLayerOptions Type of event.
    */
   sendEvent: function( dataLayerOptions ) {
-    var callback = dataLayerOptions.eventCallback;
+    const callback = dataLayerOptions.eventCallback;
     if ( Analytics.tagManagerIsLoaded ) {
       window.dataLayer.push( dataLayerOptions );
     } else if ( callback && typeof callback === 'function' ) {
-      callback();  // eslint-disable-line callback-return, inline-comments, max-len
+      callback(); // eslint-disable-line callback-return, inline-comments, max-len
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/Analytics.js]: How do you feel about moving the disable callback to the line above so we don't have to include the other two?

File: cfgov/unprocessed/js/modules/UStreamPlayer.js [hunk 3]
```diff
@@ -1,14 +1,12 @@
-'use strict';
+const VideoPlayer = require( './VideoPlayer' );
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/UStreamPlayer.js]: Outside this scope, but this shouldn't be used any more. Can you add a note to remove this file?

File: cfgov/unprocessed/js/modules/behavior/FlyoutMenu.js [hunk 4]
```diff
@@ -163,7 +164,7 @@ function FlyoutMenu( element ) { // eslint-disable-line max-statements, no-inlin
   function _handleTriggerOver() {
     if ( !_touchTriggered && !_suspended ) {
       this.dispatchEvent( 'triggerOver',
-                          { target: this, type: 'triggerOver' } );
+        { target: this, type: 'triggerOver' } );
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/behavior/FlyoutMenu.js]: Minor, should this be written as
```
this.dispatchEvent(
  'triggerOver',
  { target: this, type: 'triggerOver' }
);
```

File: cfgov/unprocessed/js/modules/behavior/FlyoutMenu.js [hunk 5]
```diff
@@ -254,9 +255,9 @@ function FlyoutMenu( element ) { // eslint-disable-line max-statements, no-inlin
       _isAnimating = true;
       _isExpanded = false;
       this.dispatchEvent( 'collapseBegin',
-                          { target: this, type: 'collapseBegin' } );
+        { target: this, type: 'collapseBegin' } );
       if ( _collapseTransitionMethod ) {
-        var hasTransition = _collapseTransition &&
+        const hasTransition = _collapseTransition &&
                             _collapseTransition.isAnimated();
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/behavior/FlyoutMenu.js]: Minor, fix left indentation (const is two chars longer than var)

File: cfgov/unprocessed/js/modules/util/breakpoint-state.js [hunk 6]
```diff
@@ -26,11 +25,11 @@ function _inBreakpointRange( breakpointRange, width ) {
...
  */
 function get( width ) {
-  var breakpointState = {};
-  var breakpointKey;
+  const breakpointState = {};
+  let breakpointKey;
   width = width || _getViewportDimensions().width;
 
-  for ( var rangeKey in _breakpointsConfig ) { // eslint-disable-line guard-for-in, no-inline-comments, max-len
+  for ( const rangeKey in _breakpointsConfig ) { // eslint-disable-line guard-for-in, no-inline-comments, max-len
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/util/breakpoint-state.js]: Same comment about not needing the inline or max-len ignores if this is moved up a line

File: cfgov/unprocessed/js/apps/financial-well-being/fwb-questions.js (hunk 1 above)
Comment (anselmbradford) [reply cfgov/unprocessed/js/apps/financial-well-being/fwb-questions.js]: It's not something that's gone now, but is not needed inside modules. [This setting](https://github.com/cfpb/cfgov-refresh/pull/3660/files#diff-1dc6ee56b778cd91e0327b52aaeaa1b9R5) can either be `module` or `script`, as module it won't include use strict inside modules.

**Comment by Scotchester** [comment]
@ans Would you call out the major rule changes that people writing JS would probably want to know as they write new JS? Things like not needing `'use strict';` in modules.

File: cfgov/unprocessed/js/modules/util/scroll.js [hunk 7]
```diff
@@ -88,17 +87,17 @@ function scrollTo( to, opts ) {
  * scrolls it into view.
  * @param {HTMLNode} elem The DOM element to check for
  * @param {Object} opts Optional parameters, including:
- * 	  offset: Distance from top of screen of element
-        when scroll is complete
+ *    offset: Distance from top of screen of element
+ * when scroll is complete
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/util/scroll.js]: Minor, fix indentation

File: cfgov/unprocessed/js/organisms/FilterableListControls.js [hunk 8]
```diff
@@ -237,26 +235,26 @@ function FilterableListControls( element ) {
...
+    let validatedField;
 
     _fieldGroups = [];
 
     fields.forEach( function loopFields( field ) {
-      var fieldIsValid = true;
-      var type = field.getAttribute( 'data-type' ) ||
+      let fieldIsValid = true;
+      const type = field.getAttribute( 'data-type' ) ||
                  field.getAttribute( 'type' ) ||
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/organisms/FilterableListControls.js]: Minor, indentation needs to be updated for const vs var char length

File: cfgov/unprocessed/js/organisms/MortgagePerformanceTrends/actions/default.js [hunk 9]
```diff
@@ -58,15 +56,15 @@ const defaultActionCreators = () => {
...
-  /**
-   * updateNational - Action dispatched when the national comparison is toggled
-   *
-   * @param {Boolean} includeComparison Include national comparison?
-   *
-   * @returns {Object} Action to include the national data in chart
-   */
+    /**
+     * updateNational - Action dispatched when the national comparison is toggled
+     *
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/organisms/MortgagePerformanceTrends/actions/default.js]: Minor: These extra line breaks aren't necessary

File: cfgov/unprocessed/js/modules/Analytics.js (hunk 2 above)
Comment (anselmbradford) [reply cfgov/unprocessed/js/modules/Analytics.js]: 👍  I like and we should do this everywhere that we have unnecessary ignore rules.

**Comment by anselmbradford** [comment]
@Scotchester Removing `use strict` from modules and using `const` and `let` are the biggies. The rest are very esoteric best practices and gotchas, like using the wrong quotes for template strings.

**Comment by anselmbradford** [comment]
@jimmynotjim Addressed your comments.

File: cfgov/unprocessed/js/modules/UStreamPlayer.js [hunk 10]
```diff
@@ -1,14 +1,14 @@
-'use strict';
+// TODO: Remove VideoPlayer module.
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/UStreamPlayer.js]: Change this to UStreamPlayer module, the VideoPlayer is still used with YouTube.

Comment (anselmbradford) [reply cfgov/unprocessed/js/modules/UStreamPlayer.js]: Oops, fixed

File: cfgov/unprocessed/js/modules/behavior/FlyoutMenu.js [hunk 11]
```diff
@@ -162,8 +168,10 @@ function FlyoutMenu( element ) { // eslint-disable-line max-statements, no-inlin
    */
   function _handleTriggerOver() {
     if ( !_touchTriggered && !_suspended ) {
-      this.dispatchEvent( 'triggerOver',
-                          { target: this, type: 'triggerOver' } );
+      this.dispatchEvent(
+        'triggerOver',
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/behavior/FlyoutMenu.js]: L:184, 196, 229, 264 all have their first param on the same line as the function and should break to a new line like this one to keep consistent.

File: cfgov/unprocessed/js/modules/transition/BaseTransition.js [hunk 12]
```diff
@@ -97,7 +96,7 @@ function BaseTransition( element, classes ) { // eslint-disable-line max-stateme
     _dom.style.oTransitionDuration = '0';
     _dom.style.transitionDuration = '0';
     _dom.removeEventListener( _transitionEndEvent,
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/transition/BaseTransition.js]: L:98 and 115 all have their first param on the same line as the function and should break to a new line like the one farther above to keep consistent.

File: cfgov/unprocessed/js/modules/transition/MoveTransition.js [hunk 13]
```diff
@@ -26,18 +24,19 @@ var CLASSES = {
...
+  const _baseTransition = new BaseTransition( element, CLASSES );
 
   /**
    * @returns {MoveTransition} An instance.
    */
   function init() {
     _baseTransition.init();
-    var _transitionCompleteBinded = fnBind( _transitionComplete, this );
+    const _transitionCompleteBinded = fnBind( _transitionComplete, this );
     _baseTransition.addEventListener( BaseTransition.END_EVENT,
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/modules/transition/MoveTransition.js]: Same comment here.

File: cfgov/unprocessed/js/organisms/MegaMenuMobile.js [hunk 14]
```diff
@@ -112,18 +109,19 @@ function MegaMenuMobile( menus ) {
       if ( menu.isExpanded() ) {
         level = _activeMenu.getData().level;
         menu.setCollapseTransition( transition,
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/organisms/MegaMenuMobile.js]: This param should also break like the others.

File: gulp/tasks/scripts.js [hunk 15]
```diff
@@ -84,7 +81,7 @@ function scriptsPolyfill() {
  */
 function scriptsModern() {
   return _processScript( webpackConfig.modernConf,
```
Comment (jimmynotjim) [inline gulp/tasks/scripts.js]: Same comment for this and all the other first params in this file.

File: test/browser_tests/cucumber/step_definitions/mega-menu.js [hunk 16]
```diff
@@ -34,118 +32,118 @@ defineSupportCode( function( { Then, When, Before } ) {
...
-                  expect( _dom.contentPolyCom.isDisplayed() )
-                  .to.eventually
-                  .equal( true )
-                )
-            );
+        .actions( )
+        .mouseMove( _dom.triggerPolyCom )
+        .perform()
+        .then( () =>
+
```
Comment (jimmynotjim) [inline test/browser_tests/cucumber/step_definitions/mega-menu.js]: This line break can be removed

File: test/browser_tests/page_objects/base-page.js [hunk 17]
```diff
@@ -20,16 +17,13 @@ class BasePage {
       return Promise.resolve();
     }
 
-    function _noOp( ) {
-
-      return;
-    }
+    function _noOp( ) { } // eslint-disable-line no-empty-function
```
Comment (jimmynotjim) [inline test/browser_tests/page_objects/base-page.js]: Minor, extra space inside empty parens and curly braces.

File: test/browser_tests/shared_objects/multi-select.js [hunk 18]
```diff
@@ -60,36 +58,36 @@ class MultiSelect {
...
 
   getDropDownCount( ) {
 
     return element.all( by.css( '.cf-multi-select .filter-match' ) )
-           .count();
+      .count();
   }
 
-  getDrodownLabelElements( ) {
+  getDropDownLabelElements( ) {
```
Comment (jimmynotjim) [inline test/browser_tests/shared_objects/multi-select.js]: Another space in empty parens.

File: test/browser_tests/shared_objects/wagtail-admin-link-chooser.js [hunk 19]
```diff
@@ -80,26 +77,26 @@ function getLinkHTML( linkAttributes, linkType = selectedLink.type ) {
...
+    } );
 }
 
 function selectPageLink( linkTitle ) {
   const linkSelector = `.modal a[data-title=${ linkTitle }]`;
   const cfgovLink = element( by.css( linkSelector ) );
 
   return cfgovLink.click()
-         .then( function( ) {
+    .then( function( ) {
```
Comment (jimmynotjim) [inline test/browser_tests/shared_objects/wagtail-admin-link-chooser.js]: Another space inside empty parens.

File: test/browser_tests/util/click-when-ready.js [hunk 20]
```diff
@@ -23,7 +21,7 @@ function clickWhenReady( elements ) {
 
     promise.then( _clickElement.bind( null, element ) )
 
-  , Promise.resolve( ) );
+    , Promise.resolve( ) );
```
Comment (jimmynotjim) [inline test/browser_tests/util/click-when-ready.js]: And another

**Comment by anselmbradford** [comment]
For documentation purposes: If you wish to see `use strict` being added despite it's removal from source files here, create a module that only has `console.log("hello world!")` and run it through webpack. It'll end up as:
```
! function(t) {
    function e(r) {
        if (n[r]) return n[r].exports;
        var o = n[r] = {
            i: r,
            l: !1,
            exports: {}
        };
        return t[r].call(o.exports, o, o.exports, e), o.l = !0, o.exports
    }
    var n = {};
    e.m = t, e.c = n, e.d = function(t, n, r) {
        e.o(t, n) || Object.defineProperty(t, n, {
            configurable: !1,
            enumerable: !0,
            get: r
        })
    }, e.n = function(t) {
        var n = t && t.__esModule ? function() {
            return t.default
        } : function() {
            return t
        };
        return e.d(n, "a", n), n
    }, e.o = function(t, e) {
        return Object.prototype.hasOwnProperty.call(t, e)
    }, e.p = "", e(e.s = 0)
}([function(t, e, n) {
    t.exports = n(1)
}, function(t, e, n) {
    "use strict";
    console.log("hello world!")
}]);
```

**Pull request merged**